From cad0484d6d0d3f90a92689d30ca29ec68b2de756 Mon Sep 17 00:00:00 2001
From: Sergey Ponomarev <stokito@gmail.com>
Date: Sun, 24 Dec 2023 19:02:26 +0200
Subject: [PATCH] cli-runopts.c: add BatchMode option

The BatchMode same as PasswordAuthentication=no but it has more priority.
So you can't re-enable the PasswordAuthentication like in:

    dbclient -v -o BatchMode=yes -o PasswordAuthentication=yes example.com

It also disables the interactive confirmation when a host key changed.

Close #224
---
 src/cli-auth.c    | 4 ++--
 src/cli-kex.c     | 7 ++++++-
 src/cli-runopts.c | 7 +++++++
 src/runopts.h     | 3 +++
 4 files changed, 18 insertions(+), 3 deletions(-)

This fixes CVE-2025-47203. Dependency Patch #8.

Index: dropbear-2017.75/cli-auth.c
===================================================================
--- dropbear-2017.75.orig/cli-auth.c
+++ dropbear-2017.75/cli-auth.c
@@ -240,7 +240,7 @@ void recv_msg_userauth_failure() {
 			}
 #endif
 #ifdef ENABLE_CLI_PASSWORD_AUTH
-			if (cli_opts.password_authentication && strncmp(AUTH_METHOD_PASSWORD, tok,
+			if (!cli_opts.batch_mode && cli_opts.password_authentication && strncmp(AUTH_METHOD_PASSWORD, tok,
 				AUTH_METHOD_PASSWORD_LEN) == 0) {
 				ses.authstate.authtypes |= AUTH_TYPE_PASSWORD;
 			}
@@ -292,7 +292,7 @@ int cli_auth_try() {
 #endif
 
 #ifdef ENABLE_CLI_PASSWORD_AUTH
-	if (!finished && cli_opts.password_authentication && (ses.authstate.authtypes & AUTH_TYPE_PASSWORD)) {
+	if (!finished && !cli_opts.batch_mode && cli_opts.password_authentication && (ses.authstate.authtypes & AUTH_TYPE_PASSWORD)) {
 		if (ses.keys->trans.algo_crypt->cipherdesc == NULL) {
 			fprintf(stderr, "Sorry, I won't let you use password auth unencrypted.\n");
 		} else {
Index: dropbear-2017.75/cli-kex.c
===================================================================
--- dropbear-2017.75.orig/cli-kex.c
+++ dropbear-2017.75/cli-kex.c
@@ -197,11 +197,16 @@ static void ask_to_confirm(unsigned char
 		m_free(fp);
 		return;
 	}
-	fprintf(stderr, "\nHost '%s' is not in the trusted hosts file.\n(%s fingerprint %s)\nDo you want to continue connecting? (y/n) ", 
+
+	fprintf(stderr, "\nHost '%s' is not in the trusted hosts file.\n(%s fingerprint %s)\n",
 			cli_opts.remotehost, 
 			algoname,
 			fp);
 	m_free(fp);
+	if (cli_opts.batch_mode) {
+		dropbear_exit("Didn't validate host key");
+	}
+	fprintf(stderr, "Do you want to continue connecting? (y/n) ");
 
 	tty = fopen(_PATH_TTY, "r");
 	if (tty) {
Index: dropbear-2017.75/cli-runopts.c
===================================================================
--- dropbear-2017.75.orig/cli-runopts.c
+++ dropbear-2017.75/cli-runopts.c
@@ -154,6 +154,7 @@ void cli_getopts(int argc, char ** argv)
 #endif
 	cli_opts.disable_trivial_auth = 0;
 	cli_opts.password_authentication = 1;
+	cli_opts.batch_mode = 0;
 #ifdef ENABLE_CLI_LOCALTCPFWD
 	cli_opts.localfwds = list_new();
 	opts.listen_fwd_all = 0;
@@ -890,6 +891,7 @@ static void add_extendedopt(const char*
 #ifdef ENABLE_CLI_ANYTCPFWD
 			"\tExitOnForwardFailure\n"
 #endif
+			"\tBatchMode\n"
 			"\tDisableTrivialAuth\n"
 			"\tPasswordAuthentication\n"
 #ifndef DISABLE_SYSLOG
@@ -920,6 +922,11 @@ static void add_extendedopt(const char*
 	}
 #endif
 
+	if (match_extendedopt(&optstr, "BatchMode") == DROPBEAR_SUCCESS) {
+		cli_opts.batch_mode = parse_flag_value(optstr);
+		return;
+	}
+
 	if (match_extendedopt(&optstr, "DisableTrivialAuth") == DROPBEAR_SUCCESS) {
 		cli_opts.disable_trivial_auth = parse_flag_value(optstr);
 		return;
Index: dropbear-2017.75/runopts.h
===================================================================
--- dropbear-2017.75.orig/runopts.h
+++ dropbear-2017.75/runopts.h
@@ -146,7 +146,10 @@ typedef struct cli_runopts {
 	int exit_on_fwd_failure;
 #endif
 	int disable_trivial_auth;
+	/** Use a password authentication or a key auth only.
+	For a BatchMode it's always -o PasswordAuthentication=no */
 	int password_authentication;
+	int batch_mode;
 #ifdef ENABLE_CLI_REMOTETCPFWD
 	m_list * remotefwds;
 #endif
